### PR TITLE
 NeMo Megatron: Add sequence parallelism and selective activation checkpointing (rebased)

### PR DESCRIPTION
**Pull request by ericharper**
# What does this PR do ?

Add sequence parallelism and selective activation checkpointing.

This PR currently depends on a fork of Apex: https://github.com/NVIDIA/apex/pull/1396

**Collection**: NLP

# Changelog 
- Update GPT config
- Update parallel_lm_logits
- Add CoreAttention
  - This is the block that gets recomputed when using selective activation checkpointing
- Update ParallelAttention to work with CoreAttention
- Update ParallelTransformerLayer
  - Layer norm calls to use sequence parallel
- Update ParallelTransformer
  - Update args
  - Update final layer norm
  - Add rng contex to forward when using sequence parallel
  - Remove transposes
- Update Pooler
  - Gather from sequence parallel regions
- Update embedding
  - Add sequence_parallel flag
  - add transpose in forward
  - scatter for dropout
- Update TransformerLanguageModel
- Update GPTModel
  - Add transposes
- All-reduce layer norm params when using sequence parallel
- Add sequence parallel arg to apex calls
- Update configs and arg checking for activation checkpointing

# Usage

```
# enable sequence parallelism
model.sequence_parallel=True

# enable selective activation checkpointing
model.activations_checkpoint_granularity=selective

# all transformer layers have CoreAttention checkpointed when using selective
model.activations_checkpoint_method=null
model.activations_checkpoint_num_layers=null
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 1]
```diff
@@ -1718,79 +2004,64 @@ def forward(
...
+                        set_inference_key_value_memory=set_inference_key_value_memory,
+                        inference_max_sequence_len=inference_max_sequence_len,
+                        rotary_pos_emb=rotary_pos_emb,
+                        position_bias=position_bias,
+                        encoder_decoder_position_bias=encoder_decoder_position_bias,
+                    )
 
         # Final layer norm.
         if self.post_process:
-            # Reverting data format change [s b h] --> [b s h].
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1792]: if we assume the order [s b h], RETRO encoder/decoder needs to be fixed as it assuems [b s h]

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 2]
```diff
@@ -1005,15 +1244,26 @@ def __init__(
 
         self.fp32_residual_connection = fp32_residual_connection  # if true move residual connections to fp32
 
+        # Layernorm on the input data.
+        if normalization == 'layernorm':
+            self.input_layernorm = get_layer_norm(
+                hidden_size, layernorm_epsilon, persist_layer_norm, sequence_parallel
+            )
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1251]: if move this out of `if self.layer_type != LayerType.retrieval_decoder_after_self_attn:` context, it will creates problem (gradient not passing to the self.input_layernorm paramemters) for the LayerType.retrieval_decoder_after_self_attn because it is going to skip the first self attention sublayer and no need of `input_layernorm`.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 3]
```diff
@@ -1038,19 +1288,15 @@ def __init__(
                 megatron_legacy=megatron_legacy,
                 bias=bias,
                 headscale=headscale,
+                activations_checkpoint_granularity=activations_checkpoint_granularity,
+                sequence_parallel=sequence_parallel,
             )
-            # Normformer normalization
```
Comment (yidong72) [inline nemo/collections/nlp/modules/common/megatron/transformer.py:1042]: Why remove the `self.post_attention_normformer_norm` ? In the forward method at L1245-L1246, it is used there.

**Review by yidong72 — CHANGES_REQUESTED** [review]
It seems my RETRO model is broken due to [b,s] order swtiched. do we have a sequence gather utility method so I can work on the full sequence of data?

File: nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py [hunk 4]
```diff
@@ -45,27 +45,35 @@ def post_language_model_processing(
...
-    output = parallel_lm_logits(lm_output, logit_weights, parallel_output)
+    output = parallel_lm_logits(lm_output, logit_weights, parallel_output, sequence_parallel=sequence_parallel)
 
     if get_key_value:
         output = [output, presents]
 
     if labels is None:
-        return output
+        # [s b h] -> [b s h]
+        return output.transpose(0, 1).contiguous()
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron/gpt_model.py:63]: Yes :(

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 5]
```diff
@@ -303,6 +314,48 @@ def optimizer_zero_grad(self, *args, **kwargs):
...
+                    grad = param.main_grad
+                else:
+                    grad = param.grad
+                grads.append(grad.data)
+        coalesced = torch._utils._flatten_dense_tensors(grads)
+        torch.distributed.all_reduce(coalesced, group=parallel_state.get_tensor_model_parallel_group())
+        for buf, synced in zip(grads, torch._utils._unflatten_dense_tensors(coalesced, grads)):
+            buf.copy_(synced)
+
+    def allreduce_gradients(self):
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:335]: Must have been added back during the rebase, I'll remove it

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:1251]: I didn't intend to move it, was just tricky to rebase. We just need it under the conditional?

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 3 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:1042]: Not intended.

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 2 above)
Comment (yidong72) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:1251]: It already has it under the condition, just remove these lines.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 6]
```diff
@@ -208,14 +211,18 @@ def training_step(self, batch, batch_idx):
         else:
             loss_mean = torch.tensor(0.0).cuda()
 
+        # when using sequence parallelism, the sequence parallel layernorm grads must be all-reduced
+        if self.cfg.get('tensor_model_parallel_size', 1) > 1 and self.cfg.get('sequence_parallel', False):
+            self.allreduce_sequence_parallel_gradients()
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:216]: > It seems my RETRO model is broken due to [b,s] order swtiched. do we have a sequence gather utility method so I can work on the full sequence of data?

Yes, there's `tensor_parallel.gather_from_sequence_parallel_region()`

File: nemo/collections/nlp/modules/common/megatron/transformer.py (hunk 2 above)
Comment (ericharper) [reply nemo/collections/nlp/modules/common/megatron/transformer.py]: Looks like this just needed to be removed.

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py [hunk 7]
```diff
@@ -232,7 +232,11 @@ def training_step(self, batch, batch_idx):
         # while async grad allreduce is enabled, bprop will keep moving forward without waiting for
         # the finish of async grad AR works. Hence, to guarantee the correctness of grads reduction,
         # we cannot start weight update until all async grad AR works are done.
-        if self.megatron_amp_o2 and self.cfg.get('pipeline_model_parallel_size', 1) == 1:
+        if (
+            self.megatron_amp_o2
+            and self.cfg.get('pipeline_model_parallel_size', 1) == 1
+            and not self.cfg.get('sequence_parallel', False)
+        ):
             torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py]: Done. Great idea!

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py:216]: > what about O1 case? does it create problem if the sequence parallel layernorm grads is all-reduced and use PLT to all reduce all the grads.

I don't think this will be an issue with O1. We still need this reduction.

**Review by MaximumEntropy — COMMENTED** [review]
Minor comment about removing async grad ar code from T5

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 8]
```diff
@@ -271,12 +274,6 @@ def training_step(self, batch, batch_idx):
             # when using pipeline parallelism, we need keep the word and position embeddings in sync
             self.allreduce_word_and_position_embeddings()
 
-        # while async grad allreduce is enabled, bprop will keep moving forward without waiting for
-        # the finish of async grad AR works. Hence, to guarantee the correctness of grads reduction,
-        # we cannot start weight update until all async grad AR works are done.
-        if self.megatron_amp_o2 and self.cfg.get('pipeline_model_parallel_size', 1) == 1:
-            torch.cuda.synchronize()
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:278]: We moved it to https://github.com/NVIDIA/NeMo/blob/9ea5d77991fe74b0fb8f863cfa583abde29df4c0/nemo/core/optim/optimizer_with_main_params.py#L415

File: nemo/collections/nlp/models/language_modeling/megatron_base_model.py [hunk 9]
```diff
@@ -104,8 +104,9 @@ def _enable_nvidia_optimizations(self):
                 NVIDIA_TORCH_MINOR = 0
 
             # Apex Persistent layer norm is supported from Nvidia PyTorch container v21.11
-            if NVIDIA_TORCH_MAJOR < 21 or (NVIDIA_TORCH_MAJOR == 21 and NVIDIA_TORCH_MINOR < 11):
-                self.cfg.persist_layer_norm = False
+            # This only depends on Apex version?
```
[the remark this answers is not in the document]
Comment (ericharper) [reply nemo/collections/nlp/models/language_modeling/megatron_base_model.py]: No. But we depend on newer containers that have this so I don't think we need the check anymore.

**Review by MaximumEntropy — APPROVED** [review]

**Review by yidong72 — APPROVED** [review]
